src/Validation/URLValidationProvider.php

```diff
@@ -0,0 +1,248 @@
+<?php
+/**
+ * Provides URL validation.
+ *
+ * @package AMP
+ * @since 2.1
+ */
+
+namespace AmpProject\AmpWP\Validation;
+
+use AMP_Validated_URL_Post_Type;
+use AMP_Validation_Error_Taxonomy;
+use AMP_Validation_Manager;
+use WP_Error;
+
+/**
+ * URLValidationProvider class.
+ *
+ * @since 2.1
+ */
+final class URLValidationProvider {
+
+	/**
+	 * Key for the transient signaling validation is locked.
+	 *
+	 * @var string
+	 */
+	const LOCK_KEY = 'amp_validation_locked';
+
+	/**
+	 * Flag to pass to get_url_validation to force revalidation.
+	 *
+	 * @param string
+	 */
+	const FLAG_FORCE_REVALIDATE = 'amp_force_revalidate';
+
+	/**
+	 * Flag to pass to get_url_validation to skip revalidation.
+	 *
+	 * @param string
+	 */
+	const FLAG_NO_REVALIDATE = 'amp_no_revalidate';
+
+	/**
+	 * The total number of validation errors, regardless of whether they were accepted.
+	 *
+	 * @var int
+	 */
+	private $total_errors = 0;
+
+	/**
+	 * The total number of unaccepted validation errors.
+	 *
+	 * If an error has been accepted in the /wp-admin validation UI,
+	 * it won't count toward this.
+	 *
+	 * @var int
+	 */
+	private $unaccepted_errors = 0;
+
+	/**
+	 * The number of URLs crawled, regardless of whether they have validation errors.
+	 *
+	 * @var int
+	 */
+	private $number_validated = 0;
+
+	/**
+	 * The validation counts by type, like template or post type.
+	 *
+	 * @var array[] {
+	 *     Validity by type.
+	 *
+	 *     @type array $type {
+	 *         @type int $valid The number of valid URLs for this type.
+	 *         @type int $total The total number of URLs for this type, valid or invalid.
+	 *     }
+	 * }
+	 */
+	public $validity_by_type = [];
+
+	/**
+	 * Locks validation.
+	 */
+	private function lock() {
+		update_option( self::LOCK_KEY, time(), false );
+	}
+
+	/**
+	 * Unlocks validation.
+	 */
+	private function unlock() {
+		delete_option( self::LOCK_KEY );
+	}
+
+	/**
+	 * Returns whether validation is currently locked.
+	 *
+	 * @return boolean
+	 */
+	public function is_locked() {
+		$lock_time = intval( get_option( self::LOCK_KEY, 0 ) );
```

Comment:
This is a matter of personal preference, but I'd say typecasting to `int` is easier to read. Typecasting is also much faster than using `intval()`, but it seems it's you only notice the difference [under heavy workload](https://stackoverflow.com/a/35949025).

/cc @schlessera @westonruter for any additional insight.

```suggestion
		$lock_time = (int) get_option( self::LOCK_KEY, 0 );
```